test/server/listener_manager_impl_test.cc

```diff
@@ -2803,6 +2803,55 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, FreebindListenerEnabled) {
   }
 }
 
+// Validate that when tcp_fast_open_queue_length is set in the Listener, we see the socket option
+// propagated to setsockopt(). This is as close to an end-to-end test as we have
+// for this feature, due to the complexity of creating an integration test
+// involving the network stack. We only test the IPv4 case here, as the logic
+// around IPv4/IPv6 handling is tested generically in
+// socket_option_impl_test.cc.
+TEST_F(ListenerManagerImplWithRealFiltersTest, FastOpenListenerEnabled) {
+  NiceMock<Api::MockOsSysCalls> os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+
+  const std::string yaml = TestEnvironment::substitute(R"EOF(
+    name: FastOpenListener
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1111 }
+    filter_chains:
+    - filters:
+    tcp_fast_open_queue_length: 1
+  )EOF",
+                                                       Network::Address::IpVersion::v4);
+  if (ENVOY_SOCKET_TCP_FASTOPEN.has_value()) {
```

Comment:
Consider refactoring with other closely related tests. The cleanest way would be to move a fair chunk of this logic into the test fixtures, `ListenerManagerImplWithRealFiltersTest`, which is the parent class effectively for reference purposes here.